src/audio/tdfb/tdfb_direction.c

```diff
@@ -0,0 +1,555 @@
+	dy = y - cd->mic_locations[mic_n].y;
+	dz = z - cd->mic_locations[mic_n].z;
+
+	/* d2 is Q8.24 meters */
+	d2 = dx * dx + dy * dy + dz * dz;
+
+	/* Squared distance is Q8.24, return Q4.12 meters */
+	d = tdfb_mic_distance_sqrt(d2);
+	return d;
+}
+
+static void theoretical_time_differences(struct tdfb_comp_data *cd, int16_t az)
+{
+	int16_t d[PLATFORM_MAX_CHANNELS];
+	int16_t src_x;
+	int16_t src_y;
+	int16_t sin_az;
+	int16_t cos_az;
+	int32_t delta_d;
+	int n_mic = cd->config->num_mic_locations;
+	int i;
+
+	sin_az = sin_fixed_16b(Q_SHIFT_LEFT((int32_t)az, 12, 28)); /* Q1.15 */
+	cos_az = cos_fixed_16b(Q_SHIFT_LEFT((int32_t)az, 12, 28)); /* Q1.15 */
+	src_x = Q_MULTSR_32X32((int32_t)cos_az, SOURCE_DISTANCE, 15, 12, 12);
+	src_y = Q_MULTSR_32X32((int32_t)sin_az, SOURCE_DISTANCE, 15, 12, 12);
+
+	for (i = 0; i < n_mic; i++)
+		d[i] = distance_from_source(cd, i, src_x, src_y, 0);
+
+	for (i = 0; i < n_mic - 1; i++) {
+		delta_d = d[i + 1] - d[0]; /* Meters Q4.12 */
+		cd->direction.timediff_iter[i] =
+			(int32_t)((((int64_t)delta_d) << 19) / SPEED_OF_SOUND);
+	}
+}
+
+static int64_t mean_square_time_difference_err(struct tdfb_comp_data *cd)
+{
+	int64_t err = 0;
+	int32_t delta;
+	int i;
+
+	for (i = 0; i < cd->config->num_mic_locations - 1; i++) {
+		delta = cd->direction.timediff[i] - cd->direction.timediff_iter[i];
+		err +=  (int64_t)delta * delta;
+	}
+
+	return err;
+}
+
+static int unwrap_radians(int radians)
+{
+	int a = radians;
+
+	if (a > PI_Q12)
+		a -= PIMUL2_Q12;
+
+	if (a < -PI_Q12)
+		a += PIMUL2_Q12;
+
+	return a;
+}
+
+static void iterate_source_angle(struct tdfb_comp_data *cd)
+{
+	int64_t err_prev;
+	int64_t err;
+	int32_t ds1;
+	int32_t ds2;
+	int az_slow;
+	int i;
+	int az_step = AZ_STEP * cd->direction.step_sign;
+	int az = cd->direction.az_slow;
+
+	/* Start next iteration opposite direction */
+	cd->direction.step_sign *= -1;
+
+	/* Get theoretical time differences for previous angle */
+	theoretical_time_differences(cd, az);
+	err_prev = mean_square_time_difference_err(cd);
+
+	for (i = 0; i < AZ_ITERATIONS; i++) {
+		az += az_step;
+		theoretical_time_differences(cd, az);
+		err = mean_square_time_difference_err(cd);
+		if (err > err_prev) {
+			az_step = -(az_step >> 1);
+			if (az_step == 0)
+				break;
+		}
+
+		err_prev = err;
+	}
+
+	az = unwrap_radians(az);
+	if (cd->direction.line_array) {
+		/* Line array azimuth angle is -90 .. +90 */
+		if (az > PIDIV2_Q12)
+			az = PI_Q12 - az;
+
+		if (az < -PIDIV2_Q12)
+			az = -PI_Q12  - az;
+	}
+
+	cd->direction.az = az;
+
+	/* Avoid low-pass filtering to zero the angle in 360 degree arrays
+	 * due to discontinuity at -180 or +180 degree point and estimation
+	 * noise. Try to camp on either side of the circle.
+	 */
+	ds1 = cd->direction.az_slow - az;
+	ds1 = ds1 * ds1;
+	ds2 = cd->direction.az_slow - (az - 360);
+	ds2 = ds2 * ds2;
+	if (ds2 < ds1)
+		az -= 360;
+
+	ds2 = cd->direction.az_slow - (az + 360);
+	ds2 = ds2 * ds2;
+	if (ds2 < ds1)
+		az += 360;
```

Comment:
something above makes me suspicious. If the first test `ds2 < ds1` was true, 360 was subtracted from `az`. And now you add 360 back to `az` for another test? An `else` missing somewhere?

Reply:
There could be an else. The second squared error eval is unnecessary if the first 360 deg addition was done. The second test is false if first was true. One multiplication saved in that case.